src/sentry/models/counter.py

```diff
@@ -30,12 +52,41 @@ class Meta:
         db_table = "sentry_projectcounter"
 
     @classmethod
-    def increment(cls, project, delta=1):
+    def increment(cls, project, delta=1) -> int:
         """Increments a counter.  This can never decrement."""
-        return increment_project_counter(project, delta)
-
-
-def increment_project_counter(project, delta=1, using="default"):
+        if features.has("projects:short-id-pre-allocation-counter", project) and delta == 1:
+            # only use the cache path if delta is 1, as in other cases we're trying to resolve
+            # a stuck counter
+            return increment_project_counter_in_cache(project)
+        else:
+            return increment_project_counter_in_database(project, delta)
+
+
+@sentry_sdk.tracing.trace
+@metrics.wraps("counter.increment_project_counter_in_cache")
+def increment_project_counter_in_cache(project, using="default") -> int:
+    redis_key = make_short_id_counter_key(project.id)
+    redis = redis_clusters.get("default")
+    short_id = redis.lpop(redis_key)
+
+    if short_id is None:  # fallback if not populated in Redis
+        metrics.incr("counter.increment_project_counter_in_cache.fallback")
+        next_id = increment_project_counter_in_database(project, using=using)
+        refill_cached_short_ids.delay(project.id, using=using)
+        return next_id
+    else:
+        metrics.incr("counter.increment_project_counter_in_cache.found_in_redis")
+        remaining = redis.llen(redis_key)
```

Comment:
It might be worth doing this as part of a pipeline so that we don't need to make two separate io calls. This will be the happy path, so it's probably worth optimizing